scoring/score.py

```diff
@@ -1,14 +1,149 @@
+import collections
+
+NUM_TOKENS_OF_EACH_COLOUR = 10
+ROBOT_MULTIPLIER = 3
+
+
 class InvalidScoresheetException(Exception):
     pass
 
 
 class Scorer(object):
+    # Note: if the arena colours (from arena.yaml) change,
+    # both this and the printed score-sheets will need to change
+    ZONE_COLOURS = [
+        'G',    # zone 0 = green
+        'O',    # zone 1 = orange
+        'P',    # zone 2 = purple
+        'Y',    # zone 3 = yellow
+    ]
+
+    ZONE_SCORE_MAP = [
+        [ 2,  2,  2,  2,  2],
+        [ 2,  7,  7,  7,  2],
+        [ 2,  7, 30,  7,  2],
+        [ 2,  7,  7,  7,  2],
+        [ 2,  2,  2,  2,  2],
+    ]
+
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
-        self._arena_data = arena_data
+        self._zone_contents = arena_data['other']['zone_contents']
+
+        for row in self._zone_contents:
+            for zone in row:
+                zone['tokens'] = collections.Counter(
+                    zone['tokens'].replace(' ', ''),
+                )
+
+        self._colour_to_team = {
+            self.ZONE_COLOURS[info['zone']]: tla
+            for tla, info in self._teams_data.items()
+        }
+
+    def validate(self, extra):
+        tokens = collections.Counter()
+        robots = collections.Counter()
+        for row in self._zone_contents:
+            for zone in row:
+                robots.update(zone['robots'])
+                tokens.update(zone['tokens'])
+
+        extra_robots = set(robots.keys()) - set(self._teams_data.keys())
+        if extra_robots:
+            raise InvalidScoresheetException(
+                "Unepxected robot(s): {!r}".format(extra_robots),
+            )
+
+        for robot, count in robots.items():
+            if count > 1:
+                raise InvalidScoresheetException(
+                    "{} appears {} times! (can only appear once)".format(
+                        robot,
+                        count,
+                    ),
+                )
+
+        extra_token_colours = set(tokens.keys()) - set(self.ZONE_COLOURS)
+        if extra_token_colours:
+            raise InvalidScoresheetException(
+                "Unepxected token colour(s): {!r}".format(extra_token_colours),
+            )
+
+        for colour, count in tokens.items():
+            if count > NUM_TOKENS_OF_EACH_COLOUR:
```

Comment:
We should probably have an accompanying exception for if there are too few tokens of a given colour?

Reply:
How do you propose that we validate that (there isn't a "tokens elsewhere" slot on the score-sheet design)?

Reply:
ah, yes, fair point :-1:

Reply:
> How do you propose that we validate that (there isn't a "tokens elsewhere" slot on the score-sheet design)?

Why did we remove this? It's a good check.

Reply:
I agree, however it seemed to be a substantial contributor to the sheets taking a long time to complete.

I also cannot recall that it has ever actually caught an issue.